### PR TITLE
MM-53467 Unrevert changes that update current user object

### PR DESCRIPTION
**Pull request by hmhealey**
#### Summary
This is blocked on https://github.com/mattermost/mattermost/pull/23912 since it reverts that while fixing the regression that it caused. Those now-unreverted fixes were about ensuring that the current user object would be updated by without refreshing (see https://mattermost.atlassian.net/browse/MM-52546).

Unfortunately, that caused the web app to lose privileged information that it had about the current user (first_name, last_name, email, auth_service) because it would be overwritten by values coming from sanitized versions of the user object. With the changes that were added, some of the ways this could happen are as simple as:

1. Having a GM channel in your LHS
2. Seeing yourself in the at mention autocomplete

We always sanitize user objects before sending them to the client, but we aren't very consistent with how we do that, so we often over-sanitize them. Typically, APIs like `getUser` return everything for the current user (minus things like the user's hashed password), almost everything for an admin (more on that below), and only the sanitized data for other, non-admin users. Some APIs will only return sanitized objects though, and those responses will be missing the privileged fields for non-admins.

Our usual method of merging objects with `{...old, ...new}` doesn't work for those fields because they're sent down to the client as `""` when sanitized out. But we also can't simply ignore empty values for those fields because they can all be empty normally. For example, you don't need to have your first/last name set in MM, and your auth_service will be set to `""` if you use email/password login.

To get around that, we can use the existence of the `notify_props` or `last_password_update` fields to tell us whether or not a user object contains privileged information. Those fields are only ever returned to the current user via APIs like `getUser`, and they're always non-zero (in Go terms), so if a user object has one of those, we can guarantee that it contains 100% correct values for those privileged fields. If we find one on the old object and not the new one when receiving data from the server, we know that we need to keep the previous value of those privileged fields.

```js
const merged = {...old, ...new};
if (old.notify_props && !new.notify_props) {
    merged.auth_service = old.auth_service;
    // and so on with the other fields
}
```

**tl;dr: We thought we could simply merge user objects together to ensure "privileged data" wasn't lost, but that didn't work because those values would be defined as empty strings which is also a valid value for them. This looks for a privileged field (notify_props) that can never be empty and uses that to help merge them manually.**

Let me know if any of that doesn't make sense. I also tried to describe that in a comment in the reducer as well so that we're less likely to accidentally break this in the future.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53467

#### Related Pull Requests
This reverts https://github.com/mattermost/mattermost/pull/23912
This re-adds https://github.com/mattermost/mattermost/pull/23219 and https://github.com/mattermost/mattermost/pull/23071

#### Release Note
```release-note
Updates the current user object more frequently without 
```

### CODE REVIEW

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts [hunk 1]
```diff
@@ -105,7 +105,7 @@ function removeProfileFromSet(state: RelationOneToMany<Team, UserProfile>, actio
 function currentUserId(state = '', action: GenericAction) {
     switch (action.type) {
     case UserTypes.RECEIVED_ME: {
-        const data = action.data || action.payload;
```
Comment (hmhealey) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts:108]: `action.payload` is a remnant of an old mobile PR that seems to have been copied around randomly throughout this file. The only actions that have a `payload` field in the web app aren't related to user objects

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts [hunk 2]
```diff
@@ -173,65 +173,81 @@ function myAudits(state = [], action: GenericAction) {
...
+    // 1. Never empty or zero by Go standards
+    // 2. Only ever sent to the current user, not even to admins, so we know that the object contains privileged data
+    //
+    // Note that admins may have the email/name/auth_service of other users loaded as well. This does not prevent that
+    // data from being replaced when merging sanitized user objects. There doesn't seem to be a way for us to detect
+    // whether the object is sanitized for admins.
+    if (existing.notify_props && (!received.notify_props || Object.keys(received.notify_props).length === 0)) {
+        merged.email = existing.email;
+        merged.first_name = existing.first_name;
+        merged.last_name = existing.last_name;
```
Comment (lieut-data) [inline webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts:214]: If `PrivacySettings.ShowEmailAddress` or `PrivacySettings.ShowFullName` are `true`, are we now discarding legitimate changes to this information?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks, @hmhealey! This makes sense, but just two comments below.

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts [hunk 3]
```diff
@@ -173,65 +173,81 @@ function myAudits(state = [], action: GenericAction) {
...
+    // Because email, first_name, last_name, and auth_service can all be empty strings regularly, we can't just
+    // merge the received user and the existing one together like we normally would. Instead, we can use the
+    // existence of existing.notify_props or existing.last_password_update to determine which object has that extra
+    // data so that it can take precedence. Those fields are:
+    // 1. Never empty or zero by Go standards
+    // 2. Only ever sent to the current user, not even to admins, so we know that the object contains privileged data
+    //
+    // Note that admins may have the email/name/auth_service of other users loaded as well. This does not prevent that
+    // data from being replaced when merging sanitized user objects. There doesn't seem to be a way for us to detect
+    // whether the object is sanitized for admins.
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts:210]: No, unfortunately not. Admins are usually sent unsanitized objects, but there's at least some channel-wide or team-wide websocket traffic that is always sent as sanitized.

The admin case is an existing issue that wasn't introduced by the reverted changes, and often admin-related code reloads the User objects manually (like in the System Console Users list), so I decided to leave it as-is. A more ideal fix for this would be to either fix all of our sanitization code to be more consistent (hard) or to add something like a field on the object for the frontend code to more clearly understand the sanitization level (less hard but still a fair bit of work)

File: webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts (hunk 2 above)
Comment (hmhealey) [reply webapp/channels/src/packages/mattermost-redux/src/reducers/entities/users.ts:214]: No, that shouldn't be possible. With the if statement above, the only case where that would be possible would be if the response or websocket event for the current user changing their name/email was sanitized. From what I can tell, that does not happen because `app.UpdateUser` uses the same mostly-unsanitized User object for both the response to the API request and for the WebSocket message that's broadcast to the affected user

**Review by yasserfaraazkhan — APPROVED** [review]
- tested on multiple browser, tab, mobile and Browser by changing user’s status
- PrivacySettings.ShowEmailAddress set to false, non-admin users should always be able to see their own email address in the profile popover.

working as expected.

will also do a post merge to cover OAuth, SAML.
